### ISSUE TITLE
Modified blacken-docs fails on some files

### ISSUE THREAD

**Issue by klieret** [hsf-training/hsf-training-matplotlib#94]
> Hi, thanks for the updates. 
> 
> there seems to be one more issue: 
> ```python
> pip install ...
>  ```
> 
> does not seem to work. I see the following error on a different file. 
> 
> command :   blacken-docs 05-mplhep.md 
>  Result  05-mplhep.md:27: code block parse error Cannot parse: 1:4: pip install mplhep
> 
> needs more fixing maybe?

_Originally posted by @amanmdesai in https://github.com/hsf-training/hsf-training-ml-webpage/issues/16#issuecomment-1197565311_

**Comment by amanmdesai** [comment]
Is there a way we can fix this in blacken-docs?

**Comment by klieret** [comment]
Ah, this is actually wrong in the lesson: `pip install ...` is not valid python. It's a bash command.

So this should be fixed by moving `pip install ...` to a `bash` code snippet like so

    ```bash
    pip install ...
    ```

**Comment by amanmdesai** [comment]
Right. I will make this change.

**Comment by klieret** [comment]
This is the full output:

```
./_episodes/05-mplhep.md:27: code block parse error Cannot parse: 1:4: pip install mplhep
./_episodes/01-introduction.md: Rewriting...
./_episodes/07-dimuonspectrum.md: Rewriting...
./_episodes/04-higgs-search.md:28: code block parse error Cannot parse: 3:0: %matplotlib inline
./_episodes/04-higgs-search.md:107: code block parse error Cannot parse: 19:0:  ~~~
./_episodes/04-higgs-search.md:147: code block parse error Cannot parse: 5:1:  <BarContainer object of 10 artists>)
./_episodes/04-higgs-search.md:177: code block parse error Cannot parse: 5:1:  <BarContainer object of 10 artists>)
./_episodes/04-higgs-search.md:238: code block parse error Cannot parse: 1:0: <Array [[1, 1, 1, 1], [1, ... 1], [1, 0, 1, 1]] type='32 * var * int32'>
./_episodes/04-higgs-search.md:251: code block parse error Cannot parse: 1:0: <Array [4, 4, 3, 4, 4, 4, ... 4, 4, 2, 2, 4, 3] type='32 * int32'>
./_episodes/04-higgs-search.md:263: code block parse error Cannot parse: 1:0: <Array [True, True, False, ... True, False] type='32 * bool'>
./_episodes/04-higgs-search.md:332: code block parse error Cannot parse: 1:13: data_A       Initial events:  32
./_episodes/04-higgs-search.md:351: code block parse error Cannot parse: 1:13: data_A       After selection:  18
./_episodes/04-higgs-search.md:377: code block parse error Cannot parse: 2:3: ~~~
./_episodes/04-higgs-search.md:415: code block parse error Cannot parse: 1:7: 302646 302646
./_episodes/04-higgs-search.md:471: code block parse error Cannot parse: 2:3: ~~~
```

The other errors are probably because there's output mixed with python code.

The only thing I don't know what to do with is the `%matplotlib inline` thing (it's not valid python, but only valid ipython). We could just move it to a general code block without language.

**Comment by amanmdesai** [comment]
Is there some way we could make blacken-docs treat %matplotlib inline as a comment?

**Comment by klieret** [comment]
I think there's the `--python-cell-magics` option to `black` (this is the tool for the actual formatting), but `blacken-docs` doesn't pass it on...

**Comment by klieret** [comment]
I think adding it as a separate cell without the `python` language specification is probably best. The line is probably separately explained anyway

**Comment by amanmdesai** [comment]
Yes. That's easier.

**Comment by amanmdesai** [comment]
> This is the full output:
> 
> ```
> ./_episodes/05-mplhep.md:27: code block parse error Cannot parse: 1:4: pip install mplhep
> ./_episodes/01-introduction.md: Rewriting...
> ./_episodes/07-dimuonspectrum.md: Rewriting...
> ./_episodes/04-higgs-search.md:28: code block parse error Cannot parse: 3:0: %matplotlib inline
> ./_episodes/04-higgs-search.md:107: code block parse error Cannot parse: 19:0:  ~~~
> ./_episodes/04-higgs-search.md:147: code block parse error Cannot parse: 5:1:  <BarContainer object of 10 artists>)
> ./_episodes/04-higgs-search.md:177: code block parse error Cannot parse: 5:1:  <BarContainer object of 10 artists>)
> ./_episodes/04-higgs-search.md:238: code block parse error Cannot parse: 1:0: <Array [[1, 1, 1, 1], [1, ... 1], [1, 0, 1, 1]] type='32 * var * int32'>
> ./_episodes/04-higgs-search.md:251: code block parse error Cannot parse: 1:0: <Array [4, 4, 3, 4, 4, 4, ... 4, 4, 2, 2, 4, 3] type='32 * int32'>
> ./_episodes/04-higgs-search.md:263: code block parse error Cannot parse: 1:0: <Array [True, True, False, ... True, False] type='32 * bool'>
> ./_episodes/04-higgs-search.md:332: code block parse error Cannot parse: 1:13: data_A       Initial events:  32
> ./_episodes/04-higgs-search.md:351: code block parse error Cannot parse: 1:13: data_A       After selection:  18
> ./_episodes/04-higgs-search.md:377: code block parse error Cannot parse: 2:3: ~~~
> ./_episodes/04-higgs-search.md:415: code block parse error Cannot parse: 1:7: 302646 302646
> ./_episodes/04-higgs-search.md:471: code block parse error Cannot parse: 2:3: ~~~
> ```
> 
> The other errors are probably because there's output mixed with python code.
> 
> The only thing I don't know what to do with is the `%matplotlib inline` thing (it's not valid python, but only valid ipython). We could just move it to a general code block without language.

Hi,
I am able to reproduce the errors, is there a fix/suggestion on how to overcome .output block? 

I tried to use 'bash' block instead of 'output', but it doesnt work. 
Thanks

**Comment by klieret** [comment]
That's definitely a bug that I introduced :(   (basically I'm matching from `~~~` to the next `~~~\n{: .language-python}` block, so it just skips the `~~~\n{: .output}`...

Minimal example to reproduce

```markdown
~~~
(array([ 3., 13.,  3.,  5.,  4.,  2.,  0.,  0.,  1.,  1.]),
 array([ 19.312525,  82.30848 , 145.30444 , 208.3004  , 271.29636 ,
        334.2923  , 397.28827 , 460.2842  , 523.28015 , 586.2761  ,
        649.2721  ], dtype=float32),
 <BarContainer object of 10 artists>)
~~~
{: .output}


~~~
fig, ax = plt.subplots()
ax.set_title("First look at samples")
ax.hist(branches["data_A"]["m4l"])
~~~
{: .language-python}
```

**Comment by amanmdesai** [comment]
I also noticed that at some places we are using
episode 1 for instance

```python `
``
whereas at others we use {: .language-python}, the example you showed for instance.

Is that something that needs a fix..i.e, making the repo more consistent?

**Comment by klieret** [comment]
Yes, this might be another way to handle this issue.

Can you check if the two markups leads to any perceivable difference in the outputs?

**Comment by amanmdesai** [comment]
Sure I will check.

**Comment by amanmdesai** [comment]
There is no observable difference in the output.

output 1 (using 3 `)
fig, ax = plt.subplots()
ax.set_title("First look at samples")
ax.hist(branches["data_A"]["m4l"])

the other output
fig, ax = plt.subplots()
ax.set_title("First look at samples")
ax.hist(branches["data_A"]["m4l"])

both are same.

**Comment by klieret** [comment]
No I meant on the rendered website. But I also think that there's probably no difference.

**Comment by amanmdesai** [comment]
Yeah I tested with the rendered website (using local build). There's no difference.

**Comment by amanmdesai** [comment]
You may check the website results here: https://amanmdesai.github.io/hsf-training-matplotlib/02-coffee-break/index.html

The markdown file where I implemented this: https://github.com/amanmdesai/hsf-training-matplotlib/blob/test-branch/_episodes/02-coffee-break.md

The python block does not have visible differences.

**Comment by amanmdesai** [comment]
> You may check the website results here: https://amanmdesai.github.io/hsf-training-matplotlib/02-coffee-break/index.html
> 
> The markdown file where I implemented this: https://github.com/amanmdesai/hsf-training-matplotlib/blob/test-branch/_episodes/02-coffee-break.md
> 
> The python block does not have visible differences.

Hi @klieret, 
 
this is just to know if the links look okay to you. 
we can then implement it on episodes.

**Comment by klieret** [comment]
Hi @amanmdesai thanks for testing this out!

Then I totally agree that we should just switch to the backticks version. It's also much easier to write. I could write a small script that converts it, or do you want to do something similar?

**Comment by amanmdesai** [comment]
Thanks, 

I will try to do it.

**Comment by amanmdesai** [comment]
One further problem--  episode 4

while python blocks are now converted in my fork -[link](https://github.com/amanmdesai/hsf-training-matplotlib/tree/episode-4-update-python-block/_episodes) 

the three backticks wont work for output block. I dont know why. Maybe this has to do with markdown or with framework?

Please check the rendered website of my fork: [link](https://amanmdesai.github.io/hsf-training-matplotlib/04-higgs-search/index.html)  

Thanks

**Comment by klieret** [comment]
Yes, but that would be ok. 

So we'd just convert the python blocks to the backticks notation and leave the output in the tilde notation

**Comment by amanmdesai** [comment]
ok. sure.

**Comment by amanmdesai** [comment]
Hi @klieret,  episode 5 doesnt give any error on using the modified blacken-docs, I think we already fixed the issue with pip install mplhep (see #95) and with #101 (for episode 4), I think this issue is resolved?

**Comment by klieret** [comment]
Great! Thanks a lot for your work! :)
